cfme/utils/template/base.py

```diff
@@ -0,0 +1,245 @@
+from contextlib import closing
+from threading import Lock
+from urllib2 import urlopen, URLError
+
+from fauxfactory import gen_alphanumeric
+
+from cfme.utils import trackerbot
+from cfme.utils.conf import cfme_data, credentials
+from cfme.utils.log import logger
+from cfme.utils.providers import get_mgmt
+from cfme.utils.ssh import SSHClient
+from cfme.utils.wait import wait_for
+
+NUM_OF_TRIES = 3
+lock = Lock()
+
+
+class TemplateUploadException(Exception):
+    """ Raised on template upload errors"""
+    pass
+
+
+def log_wrap(process_message):
+    def decorate(func):
+        def call(*args, **kwargs):
+            log_name = args[0].log_name
+            provider = args[0].provider
+            template_name = args[0].template_name
+            logger.info("(template-upload) [%s:%s:%s] BEGIN %s",
+                        log_name, provider, template_name, process_message)
+            result = func(*args, **kwargs)
+            if result:
+                logger.info("(template-upload) [%s:%s:%s] END %s",
+                            log_name, provider, template_name, process_message)
+            else:
+                logger.error("(template-upload) [%s:%s:%s] FAIL %s",
+                             log_name, provider, template_name, process_message)
+            return result
+
+        return call
+
+    return decorate
+
+
+class BaseTemplateUpload(object):
+    """ Base class for template management.
+
+    Class variables:
+        :var provider_type: type of initiated provider -- to be removed
+        :var log_name: string to be displayed in logs.
+        :var image_patters: regex to be matched when stream URL is used
+    """
+    provider_type = None
+    log_name = None
+    image_pattern = None
+
+    def __init__(self, stream=None, provider=None, template_name=None,
+                 cmd_line_args=None, **kwargs):
+        """
+        Required parameters:
```

Comment:
Conflicting docblock here, where stream, provider, and template_name have `None` defaults and thus aren't required.

Keep to google style docblocks as well, with 'Args' header